### DISCUSSION TITLE
Extending MP classes

### DISCUSSION THREAD

**Discussion by webdeb** [micropython/micropython#10811] [RP2040 / Pico] [answered]
Hi, I would like to extend the PWM object with the `slice` and `channel` attributes. It doesn't have to go into the main branch, I just would like to know how to do it in MP the right way, even if I would be locked into my own mp build.

In python itself, lets say when I create the `pwm = PWM(Pin(0))` its not possible to extend.
```
pwm = PWM(Pin(0))
pwm.slice = 0 # Error -> 'PWM' object has no attribute 'slice'
```

I tried to define the methods in the source, by creating a `slice`method. And it threw me, that this is not used. So I searched how its done with other methods, and realised, that they are coupled to the main PWM Class and did't wanted to create a mess there further.

So, any hints, how it can be done? Simple, Fast, and proper.. Even if these are three different ways :)

**Comment by cve2022** [comment]
I have a similar question, regarding the Timer class in my case.

**Comment by robert-hh** [comment, marked as answer]
The PWM class is pretty limited in methods. The class dictionary is in extmod/machine_pwm.c, and the per-port machine_pwm implementation is just included into that file. The only freedom you have is in the keyword arguments of the constructor or the init() method. The rp2 port does not support keyword arguments for the constructor, ~~but you can use pwm.init()~~, and there is no init() method. So for any extension you have to change extmod/machine_pwm.c, or extend the existing rp2 port e.g. with pwm.init().

**Reply by robert-hh** [reply]
Anyhow, in that specific case, there seems to be a fixed relation between GPIO No. and slice/channel. gpio = (slice *  2 + channel) + n * 16, where n is 0 or 1. So whenever you select a GPIO pin, the slice and channel is set. You cannot change it for a given GPIO pin, and just selecting slice and channel is not sufficient.

**Reply by webdeb** [reply]
this is totally clear, I just dont want to calculate the slice and channel each time when I just want to refer to it, because I am using a couple of pwm signals and mix them together, so I've to use some registers to change some props directly there, however, I am using the mp PWM class to create the pwm for convenience. And would like to add the slice. The demo code was just simple example of the base limitation, and not the question, if its right or not.

**Reply by robert-hh** [reply]
Did you try to create your own PWM class in python inheriting machine.PWM? There you can add any attributes you like.

**Reply by robert-hh** [reply]
Example:
```
from machine import PWM, Pin

class mypwm(PWM):
    def __init__(self, pin_no):
        self.slice = (pin_no >> 1) & 0x07
        self.channel = pin_no & 1
        super().__init__(Pin(pin_no))
```

**Reply by webdeb** [reply]
ah, thats a good one! thank you!
